### DISCUSSION TITLE
failure to connect within aws rds postgres

### DISCUSSION THREAD

**Discussion by RichardWright** [mikro-orm/mikro-orm#3462] [Q&A]
**Describe the bug**
Cannot connect to aws rds postgres db upon starting process. (Sometimes) About 5-6 hours later, it works.

**Stack trace**
```
err	{"type":"DriverException","message":"Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?","stack":"DriverException: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?\n    at PostgreSqlExceptionConverter.convertException (/usr/src/app/node_modules/.pnpm/@mikro-orm+core@5.3.1_17d14aeb6b3bf44de3c6add97b7c080a/node_modules/@mikro-orm/core/platforms/ExceptionConverter.js:8:16)\n    at PostgreSqlExceptionConverter.convertException (/usr/src/app/node_modules/.pnpm/@mikro-orm+postgresql@5.3.1_@mikro-orm+core@5.3.1/node_modules/@mikro-orm/postgresql/PostgreSqlExceptionConverter.js:42:22)\n    at PostgreSqlDriver.convertException (/usr/src/app/node_modules/.pnpm/@mikro-orm+core@5.3.1_17d14aeb6b3bf44de3c6add97b7c080a/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:187:54)\n    at /usr/src/app/node_modules/.pnpm/@mikro-orm+core@5.3.1_17d14aeb6b3bf44de3c6add97b7c080a/node_modules/@mikro-orm/core/drivers/DatabaseDriver.js:191:24\n    at async PostgreSqlDriver.find (/usr/src/app/node_modules/.pnpm/@mikro-orm+knex@5.3.1_@mikro-orm+core@5.3.1+pg@8.7.3/node_modules/@mikro-orm/knex/AbstractSqlDriver.js:49:24)\n    at async SqlEntityManager.find (/usr/src/app/node_modules/.pnpm/@mikro-orm+core@5.3.1_17d14aeb6b3bf44de3c6add97b7c080a/node_modules/@mikro-orm/core/EntityManager.js:106:25)\n    at async findWSPackets (/usr/src/app/apps/replay/dist/db/queries.js:181:31)\n    at async WebsocketEmitter.getPacketsAndSendToCallback (/usr/src/app/apps/replay/dist/server/wsEmitterClass.js:37:38)\n    at async WebsocketEmitter.getWebsocketPacketsToDispatch (/usr/src/app/apps/replay/dist/server/wsEmitterClass.js:50:13)\n    at async Timeout._onTimeout (/usr/src/app/apps/replay/dist/server/wsEmitterClass.js:22:13)\n\nprevious KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?\n    at Client_PG.acquireConnection (/usr/src/app/node_modules/.pnpm/knex@2.2.0_pg@8.7.3/node_modules/knex/lib/client.js:307:26)\n    at async Runner.ensureConnection (/usr/src/app/node_modules/.pnpm/knex@2.2.0_pg@8.7.3/node_modules/knex/lib/execution/runner.js:287:28)\n    at async Runner.run (/usr/src/app/node_modules/.pnpm/knex@2.2.0_pg@8.7.3/node_modules/knex/lib/execution/runner.js:30:19)\n    at async PostgreSqlConnection.executeQuery (/usr/src/app/node_modules/.pnpm/@mikro-orm+core@5.3.1_17d14aeb6b3bf44de3c6add97b7c080a/node_modules/@mikro-orm/core/connections/Connection.js:68:25)\n    at async PostgreSqlConnection.execute (/usr/src/app/node_modules/.pnpm/@mikro-orm+knex@5.3.1_@mikro-orm+core@5.3.1+pg@8.7.3/node_modules/@mikro-orm/knex/AbstractSqlConnection.js:88:21)\n    at async QueryBuilder.execute (/usr/src/app/node_modules/.pnpm/@mikro-orm+knex@5.3.1_@mikro-orm+core@5.3.1+pg@8.7.3/node_modules/@mikro-orm/knex/query/QueryBuilder.js:380:21)\n    at async PostgreSqlDriver.find (/usr/src/app/node_modules/.pnpm/@mikro-orm+knex@5.3.1_@mikro-orm+core@5.3.1+pg@8.7.3/node_modules/@mikro-orm/knex/AbstractSqlDriver.js:49:24)\n    at async SqlEntityManager.find (/usr/src/app/node_modules/.pnpm/@mikro-orm+core@5.3.1_17d14aeb6b3bf44de3c6add97b7c080a/node_modules/@mikro-orm/core/EntityManager.js:106:25)\n    at async findWSPackets (/usr/src/app/apps/replay/dist/db/queries.js:181:31)\n    at async WebsocketEmitter.getPacketsAndSendToCallback (/usr/src/app/apps/replay/dist/server/wsEmitterClass.js:37:38)","name":"DriverException","sql":"select \"r0\".* from \"test_recording_data\" as \"r0\" where \"r0\".\"type\" = 'ws' and \"r0\".\"heartbeat\" = false and \"r0\".\"received_at\" >= '2022-08-21T06:32:08.000Z' and \"r0\".\"received_at\" < '2022-08-21T06:32:09.000Z' order by \"r0\".\"received_at\" asc, \"r0\".\"id\" asc"}
```

**To Reproduce**

`
const commonDbOptions = {
        logger: (msg) => logger.info(msg),
        entities: [Recording, RecordingData, RecordingKeyMoment, Replay],
        forceUtcTimezone: true,
        debug: false,
        allowGlobalContext: true,
        disableIdentityMap: true,
      };

      if (options.dbDriver === 'postgres') {
        orm = await MikroORM.init<PostgreSqlDriver>({
          ...commonDbOptions,
          type: 'postgresql',
          clientUrl: process.env?.POSTGRES_CONNECTION_URL,
        });

        const em = orm.em as PostgresqlEntityManager;
        orm.em = em;
        recordingQb = em.createQueryBuilder(Recording);
        recordingDataQb = em.createQueryBuilder(RecordingData);
        driver = em.getDriver();
      }
}
`

**Expected behavior**
Fails to connect

**Additional context**

**Versions**

| Dependency | Version |
| node | 16.15
| typescript | 4.5.5 |
| mikro-orm | 5.4.0 and 5.2.1
| your-driver |  "pg": "8.8.0" (from mikro orm)

**Comment by B4nan** [comment]
> allowGlobalContext: true,

Why? This can lead to many issues. I can imagine it is causing this one as well.

> disableIdentityMap: true,

There is no such option in the ORM config.

> (Sometimes) About 5-6 hours later, it works.

Maybe the pool is actually full? Given you use the global context, I can imagine things stacking somewhere. This is not the intended usage, you should fork per request. 

Cant really help with this, especially given you didnt provide a reproduction, just one snippet that does not do anything, just initializes the ORM (and the error is not coming from ORM init, rather from usage - better say from knex, its not even from MikroORM - we manage only single instance of knex and knex is handling connection pooling).

**Reply by RichardWright** [reply]
This isn't done per request though, it's something that's on a timer that checks for records then loads them if it has responses. Should fork still be used in that instance?

**Reply by RichardWright** [reply]
I get that error straight away. the connection pool is set to 800 and the connections open are about 20 on the server.

**Reply by RichardWright** [reply]
I've taken this project on from someone else and I get the impression it's been setup incorrectly. But I can't find anywhere in the docs which is really explicit about the different ways of using mikro. 

Is it possible to use a single instance of the db connection without change tracking over the life of an application?

**Reply by B4nan** [reply]
Keep in mind it is knex responsible for connection pooling. MikroORM maintains single instance of knex. 

> Should fork still be used in that instance?

Yes, you should fork, I would say the only use case for not forking (and allowing the global context) is in (unit) tests.

> Is it possible to use a single instance of the db connection without change tracking over the life of an application?

Nope, the whole point of using this package is the change set tracking. 

I cant help you without seeing the code. You will need to provide minimal repro if you want to get some help for free... :]

**Reply by B4nan** [reply]
Btw using QB to insert/update things wont touch the context, so it should be fine. But reading via `qb.getResult()` will. You can also use `qb.execute()` that won't touch the context for read operations too.
